diff --git a/src/evdev-mt-touchpad-tap.c b/src/evdev-mt-touchpad-tap.c
index 69eecb14..ab3a5f8e 100644
--- a/src/evdev-mt-touchpad-tap.c
+++ b/src/evdev-mt-touchpad-tap.c
@@ -2453,7 +2453,7 @@ tp_tap_config_get_hold_tap_enabled(struct libinput_device *device)
 static inline enum libinput_config_hold_tap_state
 tp_hold_tap_default(struct evdev_device *device)
 {
-	return LIBINPUT_CONFIG_HOLD_TAP_DISABLED;
+	return LIBINPUT_CONFIG_HOLD_TAP_ENABLED;
 }
 
 static enum libinput_config_hold_tap_state
